File: keetanetwork-anchor-client/node-harness/src/signing.ts
/*
 * Signing interop harness.
 *
 * Wraps the anchor signing and HTTP-auth primitives so the Rust client can be
 * tested against them: signing, verification, URL/body auth, and the canonical
 * `objectToSignable` projection.
 */

import type * as SigningModule from '@keetanetwork/anchor/lib/utils/signing.js';
import type * as CommonModule from '@keetanetwork/anchor/lib/http-server/common.js';
import type * as KeetaNetModule from '@keetanetwork/keetanet-client';

import type { HarnessResponse } from './core.js';
import { referenceResolver, runHarness } from './core.js';

const refs = referenceResolver();
const signing = await refs.anchor<typeof SigningModule>('lib/utils/signing.js');
const common = await refs.anchor<typeof CommonModule>('lib/http-server/common.js');
const KeetaNet = refs.client<typeof KeetaNetModule>();
const Account = KeetaNet.lib.Account;
const Helper = KeetaNet.lib.Utils.Helper;

const signer = Account.fromSeed(Account.generateRandomSeed(), 0);
const secondary = Account.fromSeed(Account.generateRandomSeed(), 0);

/* A signable element of the harness protocol: a string, an integer, or the
 * harness's secondary account. */
interface Part {
	t: 's' | 'i' | 'a';
	v?: string | number;
}

interface SignRequest {
	cmd: 'sign';
	nonce?: string;
	timestamp?: string;
	data?: Part[];
}

interface VerifyRequest {
	cmd: 'verify';
	publicKeyAndType: string;
	nonce: string;
	timestamp: string;
	signature: string;
	data?: Part[];
}

interface ObjectToSignableRequest {
	cmd: 'objectToSignable';
	value: SigningModule.SignableInput;
}

interface AddSignatureToURLRequest {
	cmd: 'addSignatureToURL';
	baseUrl: string;
	account: string;
	nonce: string;
	timestamp: string;
	signature: string;
}

interface VerifyURLAuthRequest {
	cmd: 'verifyURLAuth';
	url: string;
	data?: Part[];
}

interface VerifyBodyAuthRequest {
	cmd: 'verifyBodyAuth';
	account: string;
	nonce: string;
	timestamp: string;
	signature: string;
	data?: Part[];
}

interface ShutdownRequest {
	cmd: 'shutdown';
}

type SigningRequest =
	SignRequest |
	VerifyRequest |
	ObjectToSignableRequest |
	AddSignatureToURLRequest |
	VerifyURLAuthRequest |
	VerifyBodyAuthRequest |
	ShutdownRequest;

function buildData(parts: Part[]): SigningModule.Signable {
	return(parts.map(function(part): string | number | typeof secondary {
		if (part.t === 's') {
			return(String(part.v));
		}
		if (part.t === 'i') {
			return(Number(part.v));
		}
		if (part.t === 'a') {
			return(secondary);
		}

		throw(new Error(`unknown signable part type: ${JSON.stringify(part)}`));
	}));
}

async function handleSign(request: SignRequest): Promise<HarnessResponse> {
	const data = buildData(request.data ?? []);
	const formatted = signing.FormatData(signer, data, request.nonce, request.timestamp);
	const signature = await signer.sign(Helper.bufferToArrayBuffer(formatted.verificationData));

	return({
		event: 'signed',
		nonce: formatted.nonce,
		timestamp: formatted.timestamp,
		verificationData: Buffer.from(formatted.verificationData).toString('hex'),
		signature: signature.getBuffer().toString('base64')
	});
}

async function handleVerify(request: VerifyRequest): Promise<HarnessResponse> {
	const data = buildData(request.data ?? []);
	const account = Account.fromPublicKeyAndType(Buffer.from(request.publicKeyAndType, 'hex'));
	const signed = { nonce: request.nonce, timestamp: request.timestamp, signature: request.signature };
	const valid = await signing.VerifySignedData(account, data, signed, { maxSkewMs: Number.MAX_SAFE_INTEGER });

	return({ event: 'verified', valid: valid });
}

function handleObjectToSignable(request: ObjectToSignableRequest): HarnessResponse {
	const signable = signing.objectToSignable(request.value);
	return({ event: 'object-to-signable', signable: signable });
}

function handleAddSignatureToURL(request: AddSignatureToURLRequest): HarnessResponse {
	const account = Account.fromPublicKeyString(request.account).assertAccount();
	const signedField = { nonce: request.nonce, timestamp: request.timestamp, signature: request.signature };
	const url = common.addSignatureToURL(request.baseUrl, { signedField, account });

	return({ event: 'signature-added', url: url.href });
}

async function handleVerifyURLAuth(request: VerifyURLAuthRequest): Promise<HarnessResponse> {
	const data = buildData(request.data ?? []);
	try {
		const account = await common.verifyURLAuth(request.url, function() { return(data); });
		return({ event: 'url-verified', valid: true, account: account.publicKeyString.get() });
	} catch {
		return({ event: 'url-verified', valid: false });
	}
}

async function handleVerifyBodyAuth(request: VerifyBodyAuthRequest): Promise<HarnessResponse> {
	const data = buildData(request.data ?? []);
	const body = {
		account: request.account,
		signed: { nonce: request.nonce, timestamp: request.timestamp, signature: request.signature }
	};

	try {
		const account = await common.verifyBodyAuth(body, function() { return(data); });
		return({ event: 'body-verified', valid: true, account: account.publicKeyString.get() });
	} catch {
		return({ event: 'body-verified', valid: false });
	}
}

async function handle(request: SigningRequest): Promise<HarnessResponse> {
	switch (request.cmd) {
		case 'sign': return(await handleSign(request));
		case 'verify': return(await handleVerify(request));
		case 'objectToSignable': return(handleObjectToSignable(request));
		case 'addSignatureToURL': return(handleAddSignatureToURL(request));
		case 'verifyURLAuth': return(await handleVerifyURLAuth(request));
		case 'verifyBodyAuth': return(await handleVerifyBodyAuth(request));
		case 'shutdown': return({ event: 'shutdown' });
	}
}

runHarness<SigningRequest>({
	event: 'ready',
	keyType: 'ECDSA_SECP256K1',
	signerPublicKeyAndType: signer.publicKeyAndType.toString('hex'),
	signerPublicKeyString: signer.publicKeyString.get(),
	secondaryPublicKeyAndType: secondary.publicKeyAndType.toString('hex')
}, handle);

File: keetanetwork-anchor-client/tests/support/signing.rs
//! Driver for the signing interop harness (`dist/signing.js`).

use keetanetwork_anchor::signing::Signed;
use serde_json::{json, Value};

use super::driver::{field_bool, field_str, verified_account, HarnessDriver, HarnessError};

/// A signature produced by the harness `FormatData`/sign path.
pub struct HarnessSignature {
	/// The DER verification bytes, hex-encoded.
	pub verification_data: String,
	/// The base64-encoded signature.
	pub signature: String,
}

/// A live signing harness driven over JSON lines.
pub struct SigningHarness {
	driver: HarnessDriver,
}

impl SigningHarness {
	/// Spawn the signing harness and wait for its `ready` line.
	pub fn start() -> Result<Self, HarnessError> {
		Ok(Self { driver: HarnessDriver::spawn("signing")? })
	}

	/// The harness-owned signer's `publicKeyAndType` hex.
	pub fn signer_public_key_and_type(&self) -> Result<&str, HarnessError> {
		self.driver.ready_field("signerPublicKeyAndType")
	}

	/// The harness-owned signer's `publicKeyString` (the `keeta_…` address used
	/// as the URL/body `account` parameter).
	pub fn signer_public_key_string(&self) -> Result<&str, HarnessError> {
		self.driver.ready_field("signerPublicKeyString")
	}

	/// The harness-owned secondary account's `publicKeyAndType` hex, used for
	/// account-typed signable parts.
	pub fn secondary_public_key_and_type(&self) -> Result<&str, HarnessError> {
		self.driver.ready_field("secondaryPublicKeyAndType")
	}

	/// Sign `data` with the harness signer for a fixed nonce and timestamp,
	/// returning the hex verification bytes and base64 signature.
	pub fn sign(&mut self, nonce: &str, timestamp: &str, data: Value) -> Result<HarnessSignature, HarnessError> {
		let request = json!({ "nonce": nonce, "timestamp": timestamp, "data": data });
		let response = self.driver.request("sign", request)?;

		let verification_data = field_str(&response, "verificationData")?.to_string();
		let signature = field_str(&response, "signature")?.to_string();
		Ok(HarnessSignature { verification_data, signature })
	}

	/// Verify a Rust-produced signature using the harness verifier.
	pub fn verify(
		&mut self,
		public_key_and_type: &str,
		nonce: &str,
		timestamp: &str,
		signature: &str,
		data: Value,
	) -> Result<bool, HarnessError> {
		let request = json!({
			"publicKeyAndType": public_key_and_type,
			"nonce": nonce,
			"timestamp": timestamp,
			"signature": signature,
			"data": data,
		});

		let response = self.driver.request("verify", request)?;
		field_bool(&response, "valid")
	}

	/// Attach a signature to `base_url` via `addSignatureToURL`, returning the
	/// resulting URL string.
	pub fn add_signature_to_url(
		&mut self,
		base_url: &str,
		account: &str,
		signed: &Signed,
	) -> Result<String, HarnessError> {
		let request = json!({
			"baseUrl": base_url,
			"account": account,
			"nonce": signed.nonce,
			"timestamp": signed.timestamp,
			"signature": signed.signature,
		});

		let response = self.driver.request("addSignatureToURL", request)?;
		Ok(field_str(&response, "url")?.to_string())
	}

	/// Verify a URL-signed request via `verifyURLAuth`, returning the
	/// authenticated account string, or `None` when rejected.
	pub fn verify_url(&mut self, url: &str, data: Value) -> Result<Option<String>, HarnessError> {
		let response = self
			.driver
			.request("verifyURLAuth", json!({ "url": url, "data": data }))?;
		verified_account(&response)
	}

	/// Verify a body-signed request via `verifyBodyAuth`, returning the
	/// authenticated account string, or `None` when rejected.
	pub fn verify_body(&mut self, account: &str, signed: &Signed, data: Value) -> Result<Option<String>, HarnessError> {
		let request = json!({
			"account": account,
			"nonce": signed.nonce,
			"timestamp": signed.timestamp,
			"signature": signed.signature,
			"data": data,
		});

		let response = self.driver.request("verifyBodyAuth", request)?;
		verified_account(&response)
	}

	/// Canonicalize `value` via `objectToSignable`, returning the resulting
	/// signable string parts.
	pub fn object_to_signable(&mut self, value: &Value) -> Result<Vec<String>, HarnessError> {
		let response = self
			.driver
			.request("objectToSignable", json!({ "value": value }))?;
		let parts = response
			.get("signable")
			.and_then(Value::as_array)
			.ok_or(HarnessError::MissingField { field: "signable" })?;

		parts
			.iter()
			.map(|part| {
				part.as_str()
					.map(str::to_string)
					.ok_or(HarnessError::MissingField { field: "signable" })
			})
			.collect()
	}

	/// Stop the harness and wait for it to exit.
	pub fn shutdown(self) -> Result<(), HarnessError> {
		self.driver.shutdown()
	}
}
